src/display/pattern_helper.js

```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
 
       var x0 = bbox[0], y0 = bbox[1], x1 = bbox[2], y1 = bbox[3];
 
-      var topLeft = [x0, y0];
-      // we want the canvas to be as large as the step size
-      var botRight = [x0 + xstep, y0 + ystep];
+      // It is a waste of memory to use the size as specified by xstep or ystep
+      // if the bounding box is smaller.
+      xstep = Math.min(xstep, x1);
```

Comment:
lots of mistakes was in the initial code, let's see if we can fix them:
- we shall rely on pattern cell width/height not on its clip boundary coordinate or xstep -- the mistake was we used xstep instead of pattern width, I don't think xstep shall participate in the calculations
- once we translated by -x0,-y0, we shall do the reverse before painting (at the scaleToContext spot)